src/platform/cmx/kernel.c

```diff
@@ -0,0 +1,271 @@
+
+	/* Validate the memory layout. */
+	assert(mem_boot_start()  >= mem_kmem_start());
+	assert(mem_kmem_start()  == mem_bootc_start());
+	assert(mem_kmem_end()    >= mem_boot_end());
+	assert(mem_utmem_start() >= mem_kmem_start());
+	assert(mem_utmem_start() >= mem_boot_end());
+	assert(mem_utmem_end()   <= mem_kmem_end());
+	assert(mem_bootc_entry() - mem_bootc_vaddr() <= mem_bootc_end() - mem_bootc_start());
+}
+
+TIM_HandleTypeDef TIM3_Handler;
+void TIM3_Init(u16 Time)
+{
+	/* General purpose timer 3 */
+	TIM3_Handler.Instance=TIM3;
+	/* Prescaler value 1024 */
+	TIM3_Handler.Init.Prescaler=1024-1;
+	/* Counting up */
+	TIM3_Handler.Init.CounterMode=TIM_COUNTERMODE_UP;
+	/* Autoreload value */
+	TIM3_Handler.Init.Period=Time;
+	/* Clock division factor - 1. This TIM2 is mounted on HCLK/2 clock tree, thus works at SYSCLK/2 speed */
+	TIM3_Handler.Init.ClockDivision=TIM_CLOCKDIVISION_DIV1;
+	/* Disable the timer */
+	HAL_TIM_Base_DeInit(&TIM3_Handler);
+	chal_timer_disable();
+	/* Low-level init and enable interrupts */
+	HAL_TIM_Base_Init(&TIM3_Handler);//, TIM_OPMODE_SINGLE);
+	/* Clear interrupt pending bit, because we used EGR to update the registers */
+	__HAL_TIM_CLEAR_IT(&TIM3_Handler, TIM_IT_UPDATE);
+	//HAL_TIM_Base_Init();
+	HAL_TIM_Base_Start_IT(&TIM3_Handler);
+}
+
+/* The low-level driver which will be called by HAL_TIM_Base_Init to set the interrupt priority */
+void HAL_TIM_Base_MspInit(TIM_HandleTypeDef *htim)
+{
+	if(htim->Instance==TIM3)
+	{
+		/* Set the interrupt priority */
+		HAL_NVIC_SetPriority(TIM3_IRQn,1,3);
+		/* Enable timer 3 interrupt */
+		HAL_NVIC_EnableIRQ(TIM3_IRQn);
+		/* Enable timer 3 clock */
+		__HAL_RCC_TIM3_CLK_ENABLE();
+	}
+}
+
+/* The low-level driver which will be called by HAL_TIM_Base_Init to set the interrupt priority */
+void HAL_TIM_Base_MspDeInit(TIM_HandleTypeDef *htim)
+{
+    if(htim->Instance==TIM3)
+	{
+		/* Set the interrupt priority */
+		//HAL_NVIC_SetPriority(TIM3_IRQn,1,3);
+		/* Enable timer 3 interrupt */
+		HAL_NVIC_DisableIRQ(TIM3_IRQn);
+	}
+}
+
+/* The actual tim3 handler */
+struct pt_regs* timer_regs;
+void tim3irqhandler(struct pt_regs* regs)
+{
+	timer_regs=regs;
+	HAL_TIM_IRQHandler(&TIM3_Handler);
+}
+
+extern int timer_process(struct pt_regs *regs);
+
+/* The call-back function */
+void HAL_TIM_PeriodElapsedCallback(TIM_HandleTypeDef *htim)
+{
+	if(htim==(&TIM3_Handler))
+	{
+		/* We are using one-shot, so disable the timer now */
+		HAL_TIM_Base_DeInit(&TIM3_Handler);
+		chal_timer_disable();
+		timer_process(timer_regs);
+	}
+}
+
+
+/* Systick is used to simulate the rdtsc. do not touch */
+void
+timer_init(void)
+{
+	HAL_SYSTICK_CLKSourceConfig(SYSTICK_CLKSOURCE_HCLK);
+	SysTick->CTRL|=SysTick_CTRL_TICKINT_Msk;
+	SysTick->LOAD=0xFFFFFF;/*1000;*/
+	SysTick->VAL=0x00;
+	NVIC_SetPriority(SysTick_IRQn,0x00);
+	SysTick->CTRL|=SysTick_CTRL_ENABLE_Msk;
+}
+
+void SysTick_Handler(void)
+{
+	rdtsc_sim++;
+}
+
+void __attribute__((optimize("O0"))) delay_us(u32 nus)
```

Comment:
Now all of these moved to HW specific libs. This is used for hardware initialization soft delays, And is hardware-specfic, now moved into hardware libs.